### PR TITLE
SGX DCAP 1.9 kernel driver and shared libraries.

### PR DESCRIPTION
**Pull request by garywest**
Build DCAP kernel driver and shared libraries.

### CODE REVIEW

File: recipes-bsp/sgx/sgx-dcap_1.9.bb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+SRC_URI  = "git://git@github.com/intel/SGXDataCenterAttestationPrimitives.git"
+SRCREV = "98976322e8b58e23256355f5cf90b9e30e37d8c1"
+
+### configure prebuilt ###
+SRC_URI += "https://download.01.org/intel-sgx/sgx-dcap/1.9/linux/prebuilt_dcap_1.9.tar.gz;name=prebuilt_dcap;subdir=${D_prebuilt_dcap}"
+SRC_URI[prebuilt_dcap.sha256sum] = "8cd0249ee49dbfd589b257cd0fa14d374d01b6c210ca7d0a14e618a48bbdb82b"
+
+configure_prebuilt () {
+    :
```
Comment (pchand20) [inline recipes-bsp/sgx/sgx-dcap_1.9.bb]: what does this override function doing here ? at least add a comment here whatever we trying to do here.

File: recipes-bsp/sgx/sgx-dcap_1.9.bb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+S_openssl="${WORKDIR}/openssl"
+File_openssl="openssl-1.1.1g"
+SRC_URI += "https://www.openssl.org/source/${File_openssl}.tar.gz;unpack=0;name=openssl;subdir=${S_openssl}"
+SRC_URI[openssl.sha256sum] = "ddb04774f1e32f0c49751e21b67216ac87852ceb056b75209af2443400636d46"
+
+configure_openssl () {
+    # Copy openssl to correct build location
+    if [ ! -f "${D_openssl}/${File_openssl}.tar.gz" ]; then
+        cp "${S_openssl}/${File_openssl}.tar.gz" "${D_openssl}"
+        if [ -d "${D_openssl}/${File_openssl}" ]; then
```
Comment (pchand20) [inline recipes-bsp/sgx/sgx-dcap_1.9.bb:57]: don't we have to remove previous version before copying tar file over ?

File: recipes-bsp/sgx/sgx_2.12.bb [hunk 3]
```diff
@@ -29,25 +29,50 @@ do_compile_prepend () {
...
+# This script causes fist boot to hang.
 pkg_postinst_ontarget_${PN} () {
-    "${sgxrootdir}/sgx_linux_x64_psw_2.12.100.3.bin"
-    rm -f "${sgxrootdir}/sgx_linux_x64_psw_2.12.100.3.bin"
+    # Verify on target
+    if [ -z "$D" ]; then
+
+        # Install SDK (optional)
+        if [ -e "${sgxrootdir}/package/sgx_linux_x64_sdk_2.12.100.3.bin" ]; then
+            "${sgxrootdir}/package/sgx_linux_x64_sdk_2.12.100.3.bin" --prefix /opt/intel < /dev/null
```
Comment (pchand20) [inline recipes-bsp/sgx/sgx_2.12.bb:51]: is this /opt/intel referring here is on yocto rootfs image ? just want to make sure its not causing harm to host system.

File: recipes-bsp/sgx/sgx-dcap_1.9.bb (hunk 2 above)
Comment (kartikeyparmar) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb:57]: the tar file is copied and old version of openssl is removed (if exist) in line #62.

Comment (jyong2) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb:57]: Please use out of tree openssl, do not bundle your own copy, the latest is already 1.1.1k.  Just make a note that this recipe is only compatible with OpenSSL 1.1.X.

File: recipes-bsp/sgx/sgx-sdk-cross_2.12.bb [hunk 4]
```diff
@@ -25,8 +25,8 @@ EXTRA_OEMAKE += "OCAMLLIB='${STAGING_LIBDIR_NATIVE}/ocaml' -C sdk signtool edger
 ### install ###
 
 do_install () {
-    install -d ${D}${sgxdirprefix}${sgxrootdir}
-    cp -r ${RECIPE_SYSROOT}${sgxsdkpath} ${D}${sgxdirprefix}${sgxsdkpath}-cross
+    ${RECIPE_SYSROOT}${sgxrootdir}/package/sgx_linux_x64_sdk_2.12.100.3.bin --prefix="${D}${sgxdirprefix}${sgxrootdir}"
```
Comment (jyong2) [inline recipes-bsp/sgx/sgx-sdk-cross_2.12.bb:29]: Please mark this recipe as only compatible with x64 build host, or modify it to use qemu user mode emulation.

File: recipes-bsp/sgx/sgx_2.12.bb (hunk 3 above)
Comment (jyong2) [reply recipes-bsp/sgx/sgx_2.12.bb:51]: https://www.yoctoproject.org/docs/latest/ref-manual/ref-manual.html#migration-2.6-post-installation-changes this seems to be the case, though I am concerned what happens on a read-only rootfs.

File: recipes-kernel/intelsgx/intelsgx.bb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+SRCREV = "98976322e8b58e23256355f5cf90b9e30e37d8c1"
+
+S = "${WORKDIR}/git/driver/linux"
+
+RPROVIDES_${PN} += "kernel-module-${PN}"
+
+### compile ###
+
+#do_compile () {
```
Comment (jyong2) [inline recipes-kernel/intelsgx/intelsgx.bb]: Are these commented lines still needed?

File: recipes-kernel/intelsgx/intelsgx.bb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+SUMMARY = "Intel(R) SGX Data Center Attestation Primitives driver"
+LICENSE = "BSD-3-Clause | GPL-2.0"
+LIC_FILES_CHKSUM = "file://License.txt;md5=633d1f5182ada7cd064194532a4a79d4"
+
+DEPENDS += " virtual/kernel"
+
+inherit module
+
+SRC_URI  = "git://git@github.com/intel/SGXDataCenterAttestationPrimitives.git"
```
Comment (jyong2) [inline recipes-kernel/intelsgx/intelsgx.bb]: Do not leave git@ username in here, add ;protocol=https.

File: recipes-bsp/sgx/sgx-dcap_1.9.bb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+# Packages required to build.
+DEPENDS += "curl sgx-sdk-cross"
+
+# Source locations
+S = "${WORKDIR}/git"
+D_prebuilt_dcap = "${S}/QuoteGeneration"
+D_sgxssl="${S}/QuoteVerification/sgxssl"
+D_openssl="${S}/QuoteVerification/sgxssl/openssl_source"
+
+SRC_URI  = "git://git@github.com/intel/SGXDataCenterAttestationPrimitives.git"
```
Comment (jyong2) [inline recipes-bsp/sgx/sgx-dcap_1.9.bb]: Do not leave git@ username in here, add ;protocol=https.

**Review by jyong2 — CHANGES_REQUESTED** [review]

**Comment by aborwank** [comment]
> are we overriding all other SRC_URI here ? if that is the motive its ok. otherwise we may need to use append construct if there something we know is already set on SRC_URI which is needed.

We are overriding the SRC_URI which build intel_sgx.ko required to be used with sgx-dcap instead of isgx.ko driver.

File: recipes-bsp/sgx/sgx-dcap_1.9.bb (hunk 1 above)
Comment (aborwank) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb]: removed this part.

File: recipes-bsp/sgx/sgx-dcap_1.9.bb (hunk 2 above)
Comment (aborwank) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb:57]: Getting errors while compiling with 1.1.1K version

File: recipes-kernel/intelsgx/intelsgx.bb (hunk 5 above)
Comment (aborwank) [reply recipes-kernel/intelsgx/intelsgx.bb]: removed

File: recipes-kernel/intelsgx/intelsgx.bb (hunk 6 above)
Comment (aborwank) [reply recipes-kernel/intelsgx/intelsgx.bb]: done

File: recipes-bsp/sgx/sgx-dcap_1.9.bb (hunk 7 above)
Comment (aborwank) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb]: done

File: recipes-bsp/sgx/sgx-dcap_1.9.bb (hunk 2 above)
Comment (aborwank) [reply recipes-bsp/sgx/sgx-dcap_1.9.bb:57]: got reply form sgx team as below:

Hi Antara,

SGX QvE is an Intel signed enclave for quote verification, it uses SgxSSL for some X.509 operations, and SgxSSL reuses/customized OpenSSL.
Here we build OpenSSL is target for SgxSSL, and the output is a trusted library for QvE.

So we cannot use system installed OpenSSL.  

Thanks,
HY

This component requires customized OpenSSL hence can not use out of tree openssl here.

File: recipes-bsp/sgx/sgx-sdk-cross_2.12.bb (hunk 4 above)
Comment (aborwank) [reply recipes-bsp/sgx/sgx-sdk-cross_2.12.bb:29]: done

**Review by aborwank — APPROVED** [review]

**Pull request merged**